### PR TITLE
feat(pageserver): support btm-gc-compaction for child branches

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

part of https://github.com/neondatabase/neon/issues/8002

## Summary of changes

For child branches, we will pull the image of the modified keys from the parant into the child branch, which creates a full history for generating key retention. If there are not enough delta keys, the image won't be wrote eventually, and we will only keep the deltas inside the child branch. We could avoid the wasteful work to pull the image from the parent if we can know the number of deltas in advance, in the future (currently we always pull image for all modified keys in the child branch)

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by skyzh** [comment]
will remove the println! before merging if I get an approval 🤣

**Comment by skyzh** [comment]
well, apart from the delta/image LSN thing, we should also figure out what does branch LSN actually mean, per https://github.com/neondatabase/neon/pull/8519/commits/6450f7453b8e8164cd025eaac33f16b459974531 changes

File: pageserver/src/tenant.rs
Comment (problame) [inline pageserver/src/tenant.rs:7780]: Let's call it `branch`?

File: pageserver/src/tenant.rs
Comment (problame) [inline pageserver/src/tenant.rs:7869]: This test would be more significant if it was testing for expected reconstruct data than expected image.

Because that would actually verify (and codify) which images from the parent are pulled up and which ones aren't.

**Review by problame — COMMENTED** [review]
If `get_ancestor_image` were to retrieve the reconstruct data instead of the image

1. we can save ourselves one walredo invocation and
2. we can seed the split bucket 0 with the reconstruct data, thereby addressing [this TODO](https://github.com/neondatabase/neon/blob/a9a8c97010b5a7bb93adadb34b5c2113bf3dfcee/pageserver/src/tenant/timeline/compaction.rs#L1202)


Further, we should invoke `get_ancestor_image` lazily.
Like, if a page has garbage in the child and that garbage has `will_init=True` in it, which I imagine to be a not too-unlikely case in practice, then we don't need the image at branch point. 


Last, the test itself is not enough to cover the behavior we're aiming for here (when we pull up an image vs when we don't). Can we somehow assert on expected reconstruct data (=sequence `Value`s) instead of the resulting image?

**Comment by skyzh** [comment]
> If get_ancestor_image were to retrieve the reconstruct data instead of the image

Yes, but it still doesn't save the I/O and memory consumption for that process, unless we include all these layers in the k-merge.

> Further, we should invoke get_ancestor_image lazily.

This is planned as a larger refactor to this function and I added it to the epic. There should be a helper function to determine whether we need the ancestor image or not.

> Can we somehow assert on expected reconstruct data (=sequence Values) instead of the resulting image?

We don't have a good facility to iterate all deltas directly in the layer map. Therefore, I will add more cases in `generate_key_retention` instead to cover more of this patch.

File: pageserver/src/tenant.rs
Comment (skyzh) [reply pageserver/src/tenant.rs]: added two test cases in generate_key_retention

File: pageserver/src/tenant.rs
Comment (skyzh) [reply pageserver/src/tenant.rs]: fixed in https://github.com/neondatabase/neon/pull/8519/commits/37525078701653f4585fcbf068aa8683ca4a428c

**Comment by skyzh** [comment]
Adding more test cases in generate_key_retention reveals the bug that the patch always pull images from the ancestor. https://github.com/neondatabase/neon/pull/8519/commits/37525078701653f4585fcbf068aa8683ca4a428c fixed the problem and adjusted related constraints around the delta layer.

**Review by problame — APPROVED** [review]
Nice test addition!

**Comment by problame** [comment]
> > If get_ancestor_image were to retrieve the reconstruct data instead of the image
> 
> Yes, but it still doesn't save the I/O and memory consumption for that process, unless we include all these layers in the k-merge.

Ok, agree that we can't do anything about this in the short term.

But let's persist this trade-off / deficiency in a comment on `get_ancestor_image` .

It's sad that we don't even use vectored get here, makes the branch btmmost compaction a lot less efficient.

**Pull request merged**
